networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,536 @@
+                        if (result[neighbor] != bestCommunity && neighbor != u) {
+                            expected = false;
+                            if (inQueue[neighbor].compare_exchange_strong(expected, true)) {
+                                newNodes.push_back(neighbor);
+                                // push new nodes to the queue in WORKING_SIZE steps
+                                if (newNodes.size() == WORKING_SIZE) {
+                                    qlock.lock();
+                                    queue.emplace(std::move(newNodes));
+                                    qlock.unlock();
+                                    // Notify threads that new is available
+                                    workAvailable.notify_all();
+                                    newNodes.clear();
+                                    newNodes.reserve(WORKING_SIZE);
+                                }
+                                assert(!expected);
+                            }
+                        }
+                    });
+                }
+            }
+
+            // queue check/wait
+            // 3 cases : newnodes not empty -> continue, newnodes empty & queue not empty ->  pop
+            // queue, both empty -> increment waiting, if waiting < #threads wait, else done notify
+            // all
+            totalNodesPerThread[omp_get_thread_num()] += currentNodes.size();
+            if (!newNodes.empty()) {
+                std::swap(currentNodes, newNodes);
+                newNodes.clear();
+                continue;
+            }
+
+            std::unique_lock<std::mutex> uniqueLock(qlock);
+            if (!queue.empty()) {
+                std::swap(currentNodes, queue.front());
+                queue.pop();
+            } else { // queue empty && newNodes empty
+                waitingForNodes++;
+                if (waitingForNodes < tcount) { // Not all nodes are done yet, wait for new work
+                    waitingForResize++;
+                    while (queue.empty() && waitingForNodes < tcount) {
+                        workAvailable.wait(uniqueLock);
+                    }
+                    if (waitingForNodes < tcount) {
+                        // Notified and not all done means there's new work
+                        std::swap(currentNodes, queue.front());
+                        queue.pop();
+                        waitingForNodes--;
+                        waitingForResize--;
+                        continue;
+                    }
+                }
+                uniqueLock.unlock(); // Notified and all done, stop.
+                workAvailable.notify_all();
+                break;
+            }
+        } while (true);
+        TRACE("Thread " + std::to_string(omp_get_thread_num()) + " worked "
+              + std::to_string(totalNodesPerThread[omp_get_thread_num()]) + "Nodes and moved "
+              + std::to_string(moved[omp_get_thread_num()]));
```

Comment:
`TRACE` accepts variadic arguments, so you can avoid manually concatenating strings by passing multiple arguments to `TRACE`. For example `TRACE("Thread ", omp_get_thread_num(), " worked ", ...) `. If you do that, you can also avoid using `std::to_string`, as this is already done automatically by `TRACE`.